R/agg_scale.R

```diff
@@ -0,0 +1,879 @@
+#' @param value_cols \[`character()`\]\cr
+#'   Value columns that should be aggregated.
+#' @param col_type \[`character(1)`\]\cr
+#'   The type of variable that is being aggregated or scaled over. Can be either
+#'   'categorical' or 'interval'.
+#' @param col_stem \[`character(1)`\]\cr
+#'   The name of the variable to be aggregated or scaled over. If aggregating an
+#'   'interval' variable should not include the '_start' or '_end' suffix.
+#' @param mapping \[`data.table()`\]\cr
+#'   For 'categorical' variables, defines how different levels of the variable
+#'   relate to each other. For aggregating 'interval' variables, it is used to
+#'   specify intervals to aggregate to, while when scaling the mapping is
+#'   inferred from the available intervals in `dt`.
+#' @param agg_function \[`function()`\]\cr
+#'   Function to use when aggregating, can be either `sum` (for counts) or
+#'   `prod` (for probabilities).
+#' @param missing_dt_severity \[`character(1)`\]\cr
+#'   How severe should the consequences of missing data that prevents
+#'   aggregation or scaling from occurring be? Can be either "stop", "warning",
+#'   "message", or "none". If not "stop", then only the possible aggregations or
+#'   scaling is done using the available data.
+#' @param drop_present_aggs \[`logical(1)`\]\cr
+#'   Whether to drop aggregates (or overlapping intervals) that are already
+#'   present in `dt` before aggregating. Default is "False" and the function
+#'   errors out.
+#'
+#' @return \[`data.table()`\] with `id_cols` and `value_cols` columns for
+#'   requested aggregates or with scaled values.
+#'
+#' @details
+#' The `agg` function can be used to aggregate over a 'categorical' variable
+#' like location, from the country level to global level or over an 'interval'
+#' variable like age from five year age-groups to all-ages combined.
+#'
+#' The `scale` function can be used to scale over a 'categorical' variable
+#' like location, so that the sub-national level aggregated together equals the
+#' national level. Similarly can be used to scale over an 'interval' variable
+#' like age so that the five year age groups aggregated together equals the
+#' all-ages value.
+#'
+#' If 'location' is the variable to be aggregated or scaled then
+#' `col_stem = 'location'` and "location" must be included in `id_cols.` If
+#' 'age' is the variable to be aggregated or scaled then `col_stem = 'age'` and
+#' 'age_start' and 'age_end' must be included in `id_cols` since both variables
+#' are needed to represent interval variables.
+#'
+#' The `mapping` argument for categorical variables must have columns called
+#' 'parent' and 'child' that represent how each possible variable relates to
+#' each other. For example if aggregating or scaling locations then mapping
+#' needs to define how each child location relates to each parent location. It
+#' is then assumed that each node in the `mapping` hierarchy will need to be
+#' aggregated to.
```

Comment:
What about multi-level aggregation/scaling?

Reply:
Multi-level mappings are fine with both functions. Just define each child to parent relationship in the mapping like in the `iran_mapping` example dataset